### PR TITLE
[onnx][importer] Add support for externalized params

### PR DESCRIPTION
**Pull request by vinayakdsci**
This patch adds support to externalize params, and store them to the given path as an IRPA file.

The IR imported with externalization should now prevent possible OOM errors happening due to large inlined parameters.

### CODE REVIEW

**Comment by vinayakdsci** [comment]
cc @kumardeepakamd.

**Comment by vinayakdsci** [comment]
There is a failure for models with dynamic dims when the parameters are externalized: https://github.com/iree-org/iree/issues/18881.

However, it appears to be unrelated to this patch.

**Review by stellaraccident — COMMENTED** [review]
Thanks. I think you have the basic idea of what to do here, but I need to take a closer look: I think there are ways to do this with less code duplication.

**Comment by vinayakdsci** [comment]
> Thanks. I think you have the basic idea of what to do here, but I need to take a closer look: I think there are ways to do this with less code duplication.

Yeah, I am sure about that too. But I had some issues with calls to super(), so I had to duplicate some code to get it working.


P.S: In case it helps, here is the IR generated by the importer for `ResNet152_vaiq_int8` https://gist.github.com/vinayakdsci/1e75b651fb66be238f4f6bd9d5e50502

**Comment by ScottTodd** [comment]
Please tag related issues on PRs (https://github.com/iree-org/iree/issues/18289 here). Also, have you read through the discussions on https://github.com/iree-org/iree/pull/18365?

(the top of https://iree.dev/developers/general/contributing/ has a note about communicating so work isn't duplicated)

**Comment by vinayakdsci** [comment]
> Also, have you read through the discussions on https://github.com/iree-org/iree/pull/18365?

Thanks, read it now.

(The changes there, though similar, seem to me to be unrelated to this PR, and it is mentioned there that this should have had been implemented separately).

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (zjgarvey) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: I wonder if there is a way to insert at the end of the symbol table?

Let me see if I understand correctly:

The referenced line inserts `global_op` into `InsertionPoint.at_block_begin(self._m.regions[0].blocks[0])`, so at the beginning of the very first block in the very first region. Is there an `InsertionPoint.at_block_end`? Or does this place it after the `func.func` ops or something?

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (zjgarvey) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: let's rename `acc` to `numel` or something more descriptive.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (zjgarvey) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: Since `max_numel` has a default value, this warning will always trigger. I think it is fine to remove this warning altogether.

Comment (zjgarvey) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: At least in my opinion, it might be more intuitive to call it `min_numel` since its the minimum number of elements required to externalize a param. Or maybe `numel_thresh`? Not sure what the best naming would be.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (zjgarvey) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: I'm a bit wary of a default `args.save_params_to` since running multiple models through the importer will replace the params file. Let's make `save_params_to` not take a default, then do something like:

```python
if args.externalize_params:
    default_param_path = Path(args.output_file).parent / Path(args.output_file).stem + "_params.irpa"
    param_path = default_param_path if args.params_save_to is None else args.params_save_to
    imp.param_archive.create_archive_file(param_path)
```

This way, the params will be located in the same directory as the output file by default, and will have the name be `<name of output file>_params.irpa`.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (zjgarvey) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py:77]: We should add a check that the provided opset version is indeed greater than the existing one. You can access the opset version from the model via `raw_model.opset_import[0].version`

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (zjgarvey) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: Yeah, I definitely think this should be renamed to `min-numel` with help description 

```python
     help="Minimum number of elements for an initializer to be externalized. Only has an effect if 'externalize-params' is true.",
```

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (vinayakdsci) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: Just passing in the block, without `at_block_end` does exactly what you are saying. It appends to the block.

That is what we do when we import other ops.

But if you want to just place the whole block at the end, I don't think it is useful. It still has to be part of the module, and we will then have to cache the globals until we emit the func.return, and then emit them in one loop.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (vinayakdsci) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: Removed the warning already. I think we should use threshold, as max or min will both be confusing.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (vinayakdsci) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: > I wonder if there is a way to insert at the end of the symbol table?

I think that is almost what it does here. The condition for the symbol table to work is that the current symbol table op should be the parent of the op you are inserting.

That is why the globals are inserted at the beginning of the first block of the first region, as that makes them a successor of the module_op, from which the symbol_table is first initialised.

**Comment by ScottTodd** [comment]
> > Also, have you read through the discussions on #18365?
> 
> Thanks, read it now.
> 
> (The changes there, though similar, seem to me to be unrelated to this PR, and it is mentioned there that this should have had been implemented separately).

Yep, I just want us to be coordinated via public GitHub issues discussions. The issue and PR I linked were tossing around the idea of replacing `iree-import-onnx` with a new API, while this PR extends the existing tool. There are architectural discussions to be had there on the issue. This component needs an owner that can manage those discussions and coordinate work between contributors/maintainers.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py:147]: Should we default this to a value like 17? We set various versions in test suites:
* https://github.com/iree-org/iree-test-suites/blob/26ee0c25a1ed1ae83aaace91a7f380f8243b4056/onnx_ops/import_onnx_tests.py#L23-L24
* https://github.com/nod-ai/SHARK-TestSuite/blob/9aa03f5f0aceac5669fff3f663b55a27490a2c15/e2eshark/onnx/operators/AveragePool/model.py#L54-L56
* https://github.com/nod-ai/SHARK-TestSuite/blob/9aa03f5f0aceac5669fff3f663b55a27490a2c15/alt_e2eshark/onnx_tests/helper_classes.py#L23-L24

(or could split the version upgrade code into its own PR, separate from the external parameters handling)

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: numel :O https://bulbapedia.bulbagarden.net/wiki/Numel_(Pok%C3%A9mon)

I'd spell out `num-elements-threshold` to be explicit... or build this into a more explicit API that allows users to control which specific parameters get externalized.
* https://iree.dev/guides/parameters/
* https://iree.dev/guides/ml-frameworks/pytorch/#using-external-parameters

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: Could also allow for configuring the parameter namespace, or at least document what the default is

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py]: I think argparse has a way to set the dtype of this to a Path, instead of treating it as a loose string

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py:77]: Might want to add logging here / for each major step

**Comment by vinayakdsci** [comment]
@ScottTodd my apologies for the delay, I have addressed all your comments in the latest commit. Also added a new flag that allows configuring the parameter scope.

**Review by zjgarvey — APPROVED** [review]
I think this is in a good spot to merge for now. @ScottTodd do you have any lingering comments?

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py:147]: Bump on this comment. Want to set a default?

> This can sometime assist with shape inference

It does more than that - many of the conversion patterns in torch-mlir set a minimum opset version far above what is possible to find in archived .onnx files (like those in https://github.com/onnx/onnx/tree/main/onnx/backend/test/data and https://github.com/onnx/models)

Take https://onnx.ai/onnx/operators/onnx__BatchNormalization.html for example - that has been defined since ONNX version `1`, but the pattern for converting it from onnx to torch specifies a minimum version of `15`: https://github.com/llvm/torch-mlir/blob/4c1518d365823c3f01d388f4d0f84112b9aa7808/lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp#L343

That's why I think we should pick a minimum version and always upgrade, or try to convert with the given version and automatically try upgrading if that fails first.

(but as stated in my first comment, the version upgrade code can be separated from the external parameter code - try to keep PRs focused on a single feature)

File: compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py:1]: Source files should have license header comments at the top

File: compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py]: Can you comment on what these caches are? What problem are they solving?

File: compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py:1]: Can you add tests for the new logic in this file?

Current in-tree `iree-import-onnx` tests:
* https://github.com/iree-org/iree/blob/main/compiler/bindings/python/test/tools/import_onnx_test.py
* https://github.com/iree-org/iree/blob/main/compiler/bindings/python/test/extras/onnx_importer_test.py

Those are run on CI with this workflow code: https://github.com/iree-org/iree/blob/f71dd1269d3ec54c07e5d77eada3f4a9ab7cb66f/.github/workflows/pkgci_unit_test.yml#L64-L70

Sample test logs: https://github.com/iree-org/iree/actions/runs/11673113040/job/32504200405#step:8:1

As needed, you can download .onnx files from public locations like what https://github.com/iree-org/iree/tree/main/compiler/bindings/python/test/build_api is doing: https://github.com/iree-org/iree/blob/f71dd1269d3ec54c07e5d77eada3f4a9ab7cb66f/compiler/bindings/python/test/build_api/mnist_builder.py#L14

File: compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py]: nit: more descriptive variable names without abbreviations please

For example:

* `initializer_name`
* `num_elements`
* `num_elements_threshold`
* `actual_symbol_name`, `vtensor_type` (`x`, `t`)

https://peps.python.org/pep-0008/#method-names-and-instance-variables

https://google.github.io/styleguide/pyguide.html#316-naming
> Function names, variable names, and filenames should be descriptive; avoid abbreviation. In particular, do not use abbreviations that are ambiguous or unfamiliar to readers outside your project, and do not abbreviate by deleting letters within a word.

> Please be mindful not to abuse single-character naming. Generally speaking, descriptiveness should be proportional to the name’s scope of visibility. For example, i might be a fine name for 5-line code block but within multiple nested scopes, it is likely too vague.

**Review by ScottTodd — COMMENTED** [review]
A few style comments and a request for test coverage. This is pretty close otherwise. Thanks!

**Comment by zjgarvey** [comment]
How big is the model.onnx file? I'd be fine with a test using small weights just to make sure the api works.

**Comment by vinayakdsci** [comment]
> How big is the model.onnx file? I'd be fine with a test using small weights just to make sure the api works.

The model file is 9.8M in size. Would you like me to make it smaller?

**Comment by zjgarvey** [comment]
Yeah, we want a functionality test, so the file size isn't necessary. Just big enough to work for the default size threshold would be ideal.

**Comment by vinayakdsci** [comment]
@zjgarvey reduced the model file size to 8.0K.

**Comment by zjgarvey** [comment]
Okay, resolve outdated conversations with @ScottTodd and we will wait for his approval.

**Comment by benvanik** [comment]
8KB is fine - thanks for making it small - git never forgets and every KB added is future pain.

File: compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py:1]: I won't have time to review this again for a few days. One thing that could be worked on in the meantime is documentation at https://iree.dev/guides/ml-frameworks/onnx/ for parameter usage, similar to https://iree.dev/guides/ml-frameworks/pytorch/#using-external-parameters. The source of that page is here: https://github.com/iree-org/iree/blob/main/docs/website/docs/guides/ml-frameworks/onnx.md

File: compiler/bindings/python/test/tools/import_onnx_test.py
Comment (ScottTodd) [inline compiler/bindings/python/test/tools/import_onnx_test.py:54]: Please move this test case under `ImportOnnxTest` in the file (just before `if __name__ == "__main__":`), so the test file starts simple and then has more complex test cases.

File: compiler/bindings/python/test/tools/import_onnx_test.py
Comment (ScottTodd) [inline compiler/bindings/python/test/tools/import_onnx_test.py:65]: This code pattern for opening a temporary file, storing the name of that file, and then later deleting it may not work on all platforms.

edit: I ran this on Windows and it worked. Phew.

**Review by ScottTodd — APPROVED** [review]
Found a bit of time to review. LGTM!

* Please move the test function down in the file
* Will you follow up with documentation, or should we file an issue?

**Comment by vinayakdsci** [comment]
@ScottTodd thanks for the approval! I have addressed the remaining comments and rebased the PR.

> * Will you follow up with documentation, or should we file an issue?

I should be able to open a PR to follow up with the documentation too.

**Comment by zjgarvey** [comment]
@vinayakdsci would you like me to merge this now?

File: compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py
Comment (ScottTodd) [inline compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py:10]: I missed this during review, but this is the only place in the compiler Python bindings that depends on the runtime Python bindings. It is used for `rt.ParameterIndex` and `create_archive_file`. That makes

```bash
pip install iree-base-compiler[onnx]
```

insufficient for running `iree-import-onnx` now, as `iree-base-runtime` must also be installed

We could

* have this check the import and fail gracefully
* add a compiler API for creating parameter archives so the runtime API is not needed
* have the compiler Python package depend on the runtime Python package

Comment (zjgarvey) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py:10]: I'm working on a few improvements in <https://github.com/iree-org/iree/pull/19217>. Will make a conditional import for this in that patch.

Comment (vinayakdsci) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py:10]: @ScottTodd I think it would be best to fail gracefully in case the import fails, at least for now. I am not very sure about how having the compiler python package depend on the runtime package should be implemented (maybe you could look at that if you find the time?), and adding a compiler API for creating parameter archives could possibly lead to code duplication IMO.

I can create a PR for the first option right away.

Comment (zjgarvey) [reply compiler/bindings/python/iree/compiler/tools/import_onnx/importer_externalization_overrides.py:10]: I think the first option, importing conditionally on whether "--externalize-parameters" is passed, with a message saying something like "externalizing parameters requires runtime api" would be the most efficient route.

**Pull request merged**
